### ISSUE TITLE
Loyalty card database

### ISSUE THREAD

**Issue by TheLastProject** [TheLastProject/Catima#89]
Stocard offers a list of cards to easily add. We should probably support something like this too:

![image](https://user-images.githubusercontent.com/1885159/98484006-d1d0b500-220c-11eb-82c1-2c4b3ea90e1a.png)

What we will need:
1. List of companies that offer loyalty cards per country and an icon
2. A way to integrate this list into the app
3. Icon support

Preferably, we have a list where every entry has an unique ID so we can just use something like a `companyId` field in the loyalty card to load the relevant icon and also update if the company ever changes. I can implement this all myself, doesn't seem that difficult, but it is very important that we don't start our own list if one already exists.

So... the important part is: does anyone already have a list of companies and logos preferably in JSON or XML format or so under an usable license, or is this something we'll need to start ourselves?

**Comment by evelikov** [comment]
This seems to overlap with https://github.com/TheLastProject/Catima/issues/8 and https://github.com/TheLastProject/Catima/issues/45, does it not? Adding cross-references for posterity ;-)

There was a question floating around wrt the legality of the icons. IMHO andOTP does it in a neat way. Namely:
 - individuals can submit MR adding/removing/updating an icon - for [example](https://github.com/andOTP/andOTP/pull/659)
 - the About page lists the app license (+link) and has a clear icon/thumbnail disclaimer

**Comment by TheLastProject** [comment]
I don't think someone else adding icons changes the legality but I am also convinced this is just fair use. A disclaimer in the about activity would be good yes

**Comment by djechelon** [comment]
>  I am also convinced this is just fair use

@TheLastProject don't. This is how I got my app removed a decade ago. (jurisdiction: Europe)

**Comment by TheLastProject** [comment]
Interesting, do you have more info about this perhaps so I could make a more educated decision?

**Comment by djechelon** [comment]
I wrote an app that allowed you to automate logging-in (yes, the concept of "free wifi" is relative) to high speed train's onboard wifi network. Basically the app automated the SMS verification.

It had a prominent logo of [one of the train companies](https://www.italotreno.it). Reviews were decent, the rail company never bothered writing me about "hey, should we discuss that you make money on a freemium app that displays our logo?" and DMCA-reported me.

Of course I am an individual running on personal project and they are a multi-million company. So Google blocked the app and demanded that the rail company allowed for unblocking. After being contacted, the company never responded to me.

So I could sue them, but the app was so cheap on budget and revenues that it wasn't worth. So I published the same app under a new package name by altering the logos to ridiculous. [One other developer](https://play.google.com/store/apps/details?id=org.paoloconte.treni_lite&hl=en&gl=US), out of all shame, was even contacted directly at the time by that rail company who **wanted to provide logo** (I had a mail conversation with that person)

**Comment by evelikov** [comment]
Fwiw I believe that the following two are indicative and clear about the goal and intend of using their data.
 - "about page has a clear icon/thumbnail disclaimer", plus
 - andOTP is free (as in beer)

Another alternative is to have a third party online icon/thumbnail database. With a once-off download, as initiated by the user.

Worst case scenario the app gets removed from Google Play, while it's still available on fdroid/elsewhere.

**Comment by IllusiveMan196** [comment]
Honestly not sure if it's worth it at all: making large enough databse takes time and one would need to include big regional/country-wide companies at least, then license images (Stocard has popularity and advertising potential behind it, which leads to deals with stores and so on; YourWallet has just some "big enough to be noticed" stores; Google Pay is everything in one and backed by... well, Google and even that does not have some stores in database), compress them while preserving readability and app size...

**Comment by mgohin** [comment]
Can this database be upgraded each time someone is adding a card and allow the app to send non private data (company name, barcode type)

**Comment by TheLastProject** [comment]
Oh, we could probably ask users upon adding an unknown store if they want to tell us about it so it can be added to the known list in a future release, yeah.

**Comment by mgohin** [comment]
With this system you don't have to wait to have a huge database to add this feature.

I don't know how it works today, but I guess you don't use any backend ? Because if you have to release too often, maybe it'll be more a pain in the a** than a great feature 😅

**Comment by TheLastProject** [comment]
There's no back-end at the moment, and that's by design. Trying to keep it so that the app does not need an internet connection at all. Sure, we can use an internet connection in a good way, but I like it when apps don't even need internet permission so you can be completely sure it will never leak any data.

**Comment by mgohin** [comment]
I like it. How could the user send data to you, a templated email ?
With some automation you could insert data from emails in your database/json

**Comment by TheLastProject** [comment]
Yeah, a templated email would be my plan. I don't mind inserting it manually from my email, that way I can also prevent people from putting in malicious strings (it's the same reason why I always manually check the translation PRs)

**Comment by mgohin** [comment]
Alright. If you add this feature with the import from image, I'd be glad to be the first to spam your inbox 😂

**Comment by jovandeginste** [comment]
Should we create a totally unaffiliated github repo with all companies, metadata and several of their logo's (in structured form), which can then be used by any other tools?

**Comment by jovandeginste** [comment]
I just found this: https://logos.fandom.com/wiki/

**Comment by TheLastProject** [comment]
I would really love a different repo yeah, so others can more easily use this in apps too and we invite more collaboration. Especially given Catima only runs on Android it would be nice to make it easier for devs for other platforms too.

**Comment by jokay** [comment]
Why not letting users add (locally stored) images themself by providing a simple image selection?

Simple and easy :wink:

**Comment by TheLastProject** [comment]
Because that doesn't properly solve things like #90. Or the lack of supporting barcode types from Fidme.

**Comment by jokay** [comment]
Correct, I thought this was for the logos only.

**Comment by Ansem93** [comment]
Out of curiosity: Does an open source/free to use database like that already exist but for other apps?
Otherwise I would be fine to have an optional template so that user can send the logo. 
Something like fixed text box with: name of the company, country and then the image.
It

**Comment by TheLastProject** [comment]
I could not find any such database, sadly. If anyone finds one I would love to know, it's a lot of work that'd be better spread over multiple people.

**Comment by Cj-Malone** [comment]
There is a similar system in the OpenStreetMap world for a database of chain stores.

We have a [git database](https://github.com/osmlab/name-suggestion-index/) that everyone can contribute to, and add brands by editing a JSON file.

The majority of the records point to a [Wikidata](https://www.wikidata.org/wiki/Wikidata:Main_Page) item. Where we can get more data about the chain, and we can often get an image via the social media profiles attached.

I think if we could redo things, we'd skip the git repo and just use Wikidata as the database.

There is already some [loyalty cards](https://w.wiki/5aoy) in Wikidata, pointing users to improve that instead of re inventing the wheel would be an amazing outcome.

**Comment by TheLastProject** [comment]
The Wikidata link looks great and is exactly the kind of thing I was looking for.

Is there a way to see all available fields? I see some like https://m.wikidata.org/wiki/Q101210200 have a logo, but I haven't found any with barcode format info yet.

**Comment by Cj-Malone** [comment]
So I'm no @wikidata expert, but. I think you need a new [property](https://www.wikidata.org/wiki/Wikidata:List_of_properties), there is a [process](https://www.wikidata.org/wiki/Wikidata:Property_creation) to make new ones.

**Comment by twig** [comment]
This is the main thing missing from Catima that I miss from Stocard.

**Comment by Vojtak42** [comment]
I'm looking for an alternative to stocard. And the feature I'm looking for is a database of logos, so I don't have to upload them myself. The app would download them on demand preferably. I didn't read this thread but it seems like no one was able to provide a good list considering the age so I think the easiest way would be if the users could upload the logos, company name and maybe barcode examples themselves to some sort of server (preferably right in the app and without any registration for the highest participation). Idk how hard that is but maybe also some sort of voting system so that when the logo gets eg 10 votes, it gets approved.
Thanks

**Comment by TheLastProject** [comment]
A system like that would be nice yeah. I'm kinda too busy to write it though, so help would be very welcome.

**Comment by ikanakova** [comment]
Or it would be nice to have it as an Aegis Authenticator with [icons](https://github.com/beemdevelopment/Aegis?tab=readme-ov-file#icon-packs).

**Comment by Ansem93** [comment]
I think a possible solution would be like Playnite on PC: it search automatically the logo of the game and add it. If it's wrong it's possible to manually search it through the program or add it manually like I did on Catima

**Comment by Vojtak42** [comment]
> I think a possible solution would be like Playnite on PC: it search automatically the logo of the game and add it. If it's wrong it's possible to manually search it through the program or add it manually like I did on Catima

Afaik Playnite grabs the data from PCGW.

**Comment by ikanakova** [comment]
In any case, I think it would be necessary to first implement SVG support (#1142) due to the size of the transmitted data, whether it would be implemented in a Playnite or Aegis way.

**Comment by SiriosDev** [comment]
> In any case, I think it would be necessary to first implement SVG support ([#1142](https://github.com/CatimaLoyalty/Android/issues/1142)) due to the size of the transmitted data, whether it would be implemented in a Playnite or Aegis way.

Hoping I have translated the discussion correctly, Why not in the `HomeAssistant Brands`  way? ([home-assistant/brands](https://github.com/home-assistant/brands)).
It seems convenient and flexible both legally and technically speaking.

**Comment by TheLastProject** [comment]
Honestly at this point I think someone should just start building something to show their way will work. We're 5 years into discussing this and not a single proof of concept exists. The main issue here I think is not technical, it is just the huge amount of work it is to collect all these logos and create some database and finding a workflow enough contributors will be OK with to maintain this. I'm already doing 95%+ of all Catima code, so I want something I can not bother with, something to leave to the community :)

If we have something usable I'd be willing to consider moving it into the CatimaLoyalty organisation if it makes sense to do so :)

**Comment by Vojtak42** [comment]
I would just like to say that if someone is willing to make the code, I will contribute the logos of the major companies in my country + those I'm using :)

**Comment by djechelon** [comment]
I just had a bright idea to share with the community (we'll talk about POC in a different moment if appreciated).
I was thinking of Whatsapp's UX for group chat logo, where you can do an internet search (Google Images?)

Actually, for Catima, I was thinking of a search to [Wikimedia Commons](https://commons.wikimedia.org), where you can find a lot of royalty-free images. Just type a company name and images should appear in the search

- Example: [Costco](https://commons.wikimedia.org/w/index.php?search=costco+logo&title=Special%3AMediaSearch&type=image)
- Example: [Esselunga](https://commons.wikimedia.org/w/index.php?search=esselunga+logo&title=Special%3AMediaSearch&type=image)
- Example: [Pret A Manger](https://commons.wikimedia.org/w/index.php?search=pret+a+manger+logo&title=Special%3AMediaSearch&type=image)

In short:
- No need to collect a (centralized) database of logos, no cost in maintaining this db, no ownership
- No need to pack logos in the apk, so no risk of DMCA report
- YMMV with the actual provider, as Wikimedia is just an example and may be not as much powerful as Google Images in returning _relevant_ results

Could the idea work?

**Comment by TheLastProject** [comment]
The fact that one of the Costco logo results in your example has a Trump look that's a hard no from me, I don't want people misunderstanding and the drama that causes 😅 We would need a trustworthy database that doesn't risk returning bad/offensive results which at worst case could get the app kicked off Google Play.

**Comment by matteosangalli** [comment]
> There is a similar system in the OpenStreetMap world for a database of chain stores.
> 
> We have a [git database](https://github.com/osmlab/name-suggestion-index/) that everyone can contribute to, and add brands by editing a JSON file.
> 
> The majority of the records point to a [Wikidata](https://www.wikidata.org/wiki/Wikidata:Main_Page) item. Where we can get more data about the chain, and we can often get an image via the social media profiles attached.
> 
> I think if we could redo things, we'd skip the git repo and just use Wikidata as the database.
> 
> There is already some [loyalty cards](https://w.wiki/5aoy) in Wikidata, pointing users to improve that instead of re inventing the wheel would be an amazing outcome.

Hello everyone, I am a bit into OpenStreetMap world and I think this is the best solution. This can provide a great database of brands with logos with correct license. Also we can offer an additional way to contribute because we share the same goal.

In a similar way [this repo](https://github.com/iptv-org/api) and [this related repo](https://github.com/iptv-org/database) do the same thing with publicly available ip-tv channels and their logos.

I'd create a database (a simple JSON file) that contains a list of brands with id, name and link to the logo. The database can be embeded into the app or hosted on a repo. The database can be created manually by inserting the most popular brands or it can be created automatically by using some query on Wikidata ans NSI. The database can contain also other data such as
- the country where the brand is present (available on NSI or Wikidata) to let the user see only the brands "close" to him, maybe based on the localization preferences.
- a "primary color" to color the background of the card

> Yeah, a templated email would be my plan. I don't mind inserting it manually from my email, that way I can also prevent people from putting in malicious strings (it's the same reason why I always manually check the translation PRs)

I think that the database can also be not very complete at start but it can be improved with time by providing to users an easy way to report new brands. This email approch should work.

I'll try to make some code to implement this
